File: hard/Best Time to Buy and Sell Stock III.js
/*
    思路 : 
        買賣股票的第三題 , 這一題增加的限制為我們總共只能進行2次交易 , 
        更加general來說就是得要計算總計k次交易可以得到的最大金錢 


        根據股票問題一貫思路的DP定義 : 
        可以設定狀態包含了 :
        1. 天數 
        2. 可以交易的最大次數 k 
        3. 是否持有股票 

        dp[i][k][j]  : 在第i天 , 最多可以交易K次的情況下 , 是否持有股票可得的最大收益 
        我們要求的就是在最後一天手上沒有股票時的最大收益(必定大於等於拿著股票) 

        此題的狀態比較多且狀態轉移複雜 , 從base case先看 

        # 第一天 ,利益只會是0或你買進了一次
        dp[0][:][0] = 0 
        dp[0][:][1] = -price[0] 

        # 最大可以交易次數為0 , 那收益也會是0 
        dp[:][0][:] = 0 

        -- 狀態轉移  : 

        這一題的狀態轉移 , 我們在每一個狀態下都有幾個動作 : 賣出 , 買入 , 不動 

        # 如果我們手上沒有股票 : 那代表我們這一回合賣掉了 , 或著上回合就沒有
        dp[i][K][0] = max (  price[i] + dp[i-1][K][1]  , dp[i-1][K][0]   )
        
        # 如果手上有股票 : 則代表應該是這一回合買入 , 或著上回合就有

        # --- !!! 特別要注意針對交易次數的限制 , 因為一次完整的買入賣出才算交易 , 我們在買入的時候才更新K 
        交易是從buy開始的 , 因此不能在sell的時候更新K , 

        dp[i][K][1] = max ( dp[i-1][K-1][0] - price[i] , dp[i-1][K][1]     )

        // !! 我在實做的時候犯了一個錯 
        就是把K值設置為2 (這一題限制交易兩次 , 但實際上狀態為 交易0次,1次,2次) 
    
*/


/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    
    let K = 2+1 ; 
    let dp = Array.from( new Array(prices.length) , (v)=> Array.from(
        new Array(K) , (v)=> new Array(2).fill(0)  
        ) );

    // 更新base case : 
    for (let k = 0 ; k < K ; k++){
        dp[0][k][0] = 0 ; 
        dp[0][k][1] = -prices[0]
    }
    for (let i = 0 ; i<prices.length ; i++){
        dp[i][0][0] = 0;
        dp[i][0][1] = 0 ;
    }

    // 在我們的狀態轉移過程 , 計算dp[i][k][j] , 會需要前一天的i-1, k-1計算完成
    // 而在已知Base case的情況下 , 我們的這兩個狀態從1開始推進就好


    // 推進天數狀態
    for (let i = 1 ; i < prices.length ; i++){
        // 推進K 狀態
        for (let k = 1 ; k < K ; k++){

            // 沒有持股 : 賣出 or 本就沒有 
            dp[i][k][0] = Math.max( prices[i] + dp[i-1][k][1] , dp[i-1][k][0] )

            // 有持股 : 買入 or 本就有 
            dp[i][k][1] = Math.max( dp[i-1][k-1][0] - prices[i] , dp[i-1][k][1] )

        }
    }

    // 最終解在 dp[price.length-1][K-1][0]
    // 代表在到達最後一天時 , 在題目給定的K數下可以得到的最大利潤

    console.log(dp[prices.length-1]);

    return dp[prices.length-1][K-1][0] ; 

};

File: TOP150/Best Time to Buy and Sell Stock III.py
from typing import List 
from numpy import array
class Solution:

    def maxProfit(self, prices: List[int]) -> int:
        
        # 最多只能執行k次的buy-sell 
        # if k == 1 dp solution  : dp[i] = ( max profit if keep stock in day i , max profit if non stock in day i ) 
        # generalize to k > 1 :
        # dp[i][k] = (  max profit if keep stock and make k time transation   , max profit if non stock and k time transation)
        """ 
        base case 
        1.dp[:][0] = (  0 ,  0 )   
        2.dp[0][1] = ( -prices[0] , 0  ) # -> 這邊有錯誤 , 應該是dp[0][:] = (-prices[0],0) 
        
        state transition : 
            dp[i][k] =  (     
                max ( dp[i-1][k-1][1] - prices[i]  , dp[i-1][k][0]
                max ( dp[i-1][k][0] + prices[i]    , dp[i-1][k][1]   )
        )

        第i天 , 並在執行了k次交易後(以購買算做一次) . 
        則拿著股票的最大收益值 會等於 前一天就已經買了(同樣在k次)  或著今天買 , 則收益為前一天在少一次交易的情況下沒有股票的最大值-prices[i]
        而沒拿股票的最大收益 , 則是 前一天有拿著股票的最大收益加上今天賣出 , 或著 前一天就沒拿著
        
        """
        
        # maximum transation times 
        K = 2 
        dp = [ [ (0,0) for i in range(K+1) ] for j in range(len(prices))  ] 

        # base case : 
        # dp[0][1] = (-prices[0] , 0)
        for k in range(K+1):
            dp[0][k] = (-prices[0] , 0)
        
        # 推進DP狀態 
        
        for i in range( 1 , len(prices) ): 
            
            for k in range( 1,K+1 ): 
                dp[i][k] = (
                    # 在前一天少一次交易狀況下的無股最大 - 買股價格 , 前一天就買完了 
                    max( dp[i-1][k-1][1] - prices[i]  ,  dp[i-1][k][0] ), 
                    
                    # 賣掉今天的股票 , 收益為昨天買入的加上今天賣出 , 或著前一天就買入的繼續拿著
                    max( dp[i-1][k][0] + prices[i] , dp[i-1][k][1]      )
                    
                )
        # 最終答案位在 dp[len(prices)-1][k] 
        return dp[len(prices)-1][k][1]
    
# test = [3,3,5,0,0,3,1,4]
test = [1,2,3,4,5]
S = Solution()
print(S.maxProfit(test))